examples/paho-mqtt/Makefile.board.dep

```diff
@@ -0,0 +1,10 @@
+# Put board specific dependencies here
+
+ifneq (,$(filter arch_esp32,$(FEATURES_USED)))
+  CFLAGS += -DESP_WIFI_SSID=\"define_yours\" -DESP_WIFI_PASS=\"secret\"
```

Comment:
```suggestion
```
These settings shouldn't be made in the Makefile, otherwise they can't be set without changing the Makefile.

Reply:
Where do you think they should go?

Reply:
I think he means something like we have in `examples/gnrc_border_router/Makefile.wifi.conf`. You can put this into `examples/paho-mqtt/Makefile` directly. The border router has many use cases, that's why it is in a separate, include-able Makefile.

Reply:
Exactly, it allows to configure WiFi settings by environment variables at command line without the necessity to change the Makefile.

Reply:
```
WIFI_SSID ?= "Your_WiFi_name"
WIFI_PASS ?= "Your_secure_password"

ifneq (,$(filter arch_esp32,$(FEATURES_USED)))
  CFLAGS += -DESP_WIFI_SSID=\"$(WIFI_SSID)\"
  CFLAGS += -DESP_WIFI_PASS=\"$(WIFI_PASS)\"
endif
```

Reply:
(`CFLAGS` extensions don't interfere with the dependency resolution of the build system, so it is ok, to put it into the application's main `Makefile`; sorry I was confused about that earlier)

Reply:
Could you explain this @miri64 ? when you talk about dependency resolution? Maybe to give an example?

Reply:
I hardly understand it myself, @aabadie can explain it better, I think, but I can try.

As you might have noticed, we resolve dependencies of modules using various `Makefile*.dep` files. In contrast to the main `Makefile` of an application, module or package those are included later in the configuration gathering phase, when things board-specific things like MCU, features of the board, etc. are clear. This is why board-specific module dependencies need to go into a separate file, otherwise commands like `make info-boards-supported` (which is e.g. used by our CI to determine if it can build the application) won't work, since it will always assume the `BOARD` you define in the application for the `ifeq (<board>,$(BOARD)))` stuff.

BTW: order of evaluation is still important so

```Makefile
ifneq (,$(filter arch_esp32,$(FEATURES_USED)))
  CFLAGS += -DESP_WIFI_SSID=\"$(WIFI_SSID)\"
  CFLAGS += -DESP_WIFI_PASS=\"$(WIFI_PASS)\"
endif
```

Needs to go _after_ the `include $(RIOTBASE)/Makefile.include` in your application, since the `$(FEATURES_USED)` list is only filled in their based on the board you selected.

Reply:
For your purposes its only important to know, that the WiFi config stuff belongs in the main `Makefile` but the

```Makefile
ifneq (,$(filter arch_esp32,$(FEATURES_USED)))
  USEMODULE += esp_wifi
endif

ifneq ($(BOARD),native)
  USEMODULE += netdev_default
endif
```

belongs to `Makefile.board.dep`, as it changes `USEMODULE` based on board-specific configuration.

Reply:
mmm not sure that I understood everything, but give me some time. Thank you anyways :)